File: src/Tree/BinaryTreePostorderTraversal/BinaryTreePostorderTraversal_iterative.java
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        //Base case
        if(root == null) return res;
        
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root, pre = null;
        
        while(cur != null || !stack.isEmpty()){
            if(cur != null){
                stack.add(cur);
                cur = cur.left;
            }else{
                TreeNode top = stack.peek();
                if(top.right == null || top.right == pre){
                    res.add(top.val);
                    stack.pop();
                    pre = top;
                    cur = null; //Don't traverse the left again
                }else{
                    cur = top.right;
                }
            }
        }
        return res;
    }
}

File: src/Tree/BinaryTreePostorderTraversal/BinaryTreePostorderTraversal_iterative.py
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        
        if root is None:
            return res
        
        stack = []
        
        cur = root
        pre = None
        
        while cur is not None or len(stack) != 0:
            if cur is not None:
                stack.append(cur)
                cur = cur.left
            else:
                top = stack[-1] # peek the last element
                if top.right is None or top.right is pre:
                    res.append(top.val)
                    stack.pop()
                    pre = top
                    cur = None # Don't traverse the left again
                else:
                    cur = top.right #Ready to traverse the right subtree
                    
        return res
        
